src/drivers/intel/cavs/hda-dma.c

```diff
@@ -612,6 +612,9 @@ static int hda_dma_probe(struct dma *dma)
 	int i;
 	struct hda_chan_data *chan;
 
+	if (dma_get_drvdata(dma))
```

Comment:
What's the use case here ? I guess topology could create the device twice in error ?

Reply:
(also applies to the multiple probe question) my assumptions:
- some devices may be created on init, some on demand, the client does not need to know whether the probe() must be called before get() or not, otherwise may try to probe the device that is already created.
- there is no api to 'return' the device back to the pool (calling remove() implicitly inside for some of them) yet, so get-probe sequences may be called again by a destroyed and then re-created topology component.

Reply:
Actually the get() could check whether the device exists before calling probe() and the get_drvdata() test could stay there but result in error (probe had not been protected against double allocation before I guess).

Reply:
ok thanks, I see - I think the simplest approach for us would be to only probe() devices when added by topology and remove() them when topology is unloaded.  This probably can be summarized with some rules (that we could also document).

1)  we dont automatically probe() any pipeline device unless its's added by topology.
2) we only remove component device when it's unloaded by topology.
3) component added twice would return an error.
4) active component removal would return an error. 
5) pipeline trigger start will fail if pipeline has no valid sink and source endpoints.

The driver would manage the component load/unload and ref count components in pipelines that shared components. @ranj063 is working on driver code for this shortly (PM related)

The thing we have to watch for is non pipeline IO drivers like low level DMACs, SSP, these would need reference counted by host.c and dai.c respectively.

Reply:
@lgirdwood Agree. However (3) & (4) in the FW would require internal reference counting as well since dmacs are shared between components, ssp dais are also shared by the Rx and Tx paths (Btw since probe() is unprotected yet, there is an existing mem leak here)

In both, dmac and ssp cases devices are acquired by independent calls through the infrastructure stack. So multiple calls to dai_get() and dma_get() for the same device are valid. So either 

(a) there is internal ref counting inside the fw to know when to call probe()/remove(), no counting in the driver, doubling this seems to be not necessary,

(b) driver explicitly creates/destroys the devices with dedicated IPC calls and fw is simple (or flags added to the existing topology calls).

What is your recommendation?

Reply:
@lgirdwood Please take a look at the latest implementation that handles ref counting (simple one, protected by early initialized locks) in the lib layer. Done for the get() side, to be completed on the put() side if this is acceptable direction.